hi-ml/src/health_ml/utils/regression_test_utils.py

```diff
@@ -0,0 +1,180 @@
+    If 0.0, do not allow any discrepancy.
+    :return: An empty string if the files appear identical, or otherwise a brief error message. If there is a
+    mismatch, details about the mismatch are printed via logging.warning.
+    """
+    def print_lines(prefix: str, lines: List[str]) -> None:
+        num_lines = len(lines)
+        count = min(5, num_lines)
+        logging.warning(f"{prefix} {num_lines} lines, first {count} of those:")
+        logging.warning(os.linesep.join(lines[:count]))
+
+    def try_read_csv(prefix: str, file: Path) -> Optional[pd.DataFrame]:
+        try:
+            return pd.read_csv(file)
+        except Exception as ex:
+            logging.warning(f"{prefix} file can't be read as CSV: {str(ex)}")
+            return None
+
+    if expected.suffix == CSV_SUFFIX:
+        expected_df = try_read_csv("Expected", expected)
+        actual_df = try_read_csv("Actual", actual)
+        if expected_df is None or actual_df is None:
+            return FILE_FORMAT_ERROR
+        try:
+            pd.testing.assert_frame_equal(actual_df, expected_df, rtol=csv_relative_tolerance)
+        except Exception as ex:
+            logging.warning(str(ex))
+            return CONTENTS_MISMATCH
+    elif expected.suffix in TEXT_FILE_SUFFIXES:
+        # Compare line-by-line to avoid issues with line separators
+        expected_lines = expected.read_text().splitlines()
+        actual_lines = actual.read_text().splitlines()
+        if expected_lines != actual_lines:
+            print_lines("Expected", expected_lines)
+            print_lines("Actual", actual_lines)
+            return CONTENTS_MISMATCH
+    else:
+        expected_binary = expected.read_bytes()
+        actual_binary = actual.read_bytes()
+        if expected_binary != actual_binary:
+            logging.warning(f"Expected {len(expected_binary)} bytes, actual {len(actual_binary)} bytes")
+            return CONTENTS_MISMATCH
+    return ""
+
+
+def compare_folder_contents(
+    expected_folder: Path,
+    actual_folder: Optional[Path] = None,
+    run: Optional[Run] = None,
+    csv_relative_tolerance: float = 0.0,
+) -> List[str]:
+    """
+    Compares a set of files in a folder, against files in either the other folder or files stored in the given
+    AzureML run. Each file that is present in the "expected" folder must be also present in the "actual" folder
+    (or the AzureML run), with exactly the same contents, in the same folder structure.
+    For example, if there is a file "<expected>/foo/bar/contents.txt", then there must also be a file
+    "<actual>/foo/bar/contents.txt". If `actual_folder` is provided, then this is used to compare files against the
+    set file files in `expected_folder`, irrespective of the value in `run`. If `run` is provided, the files
+    uploaded to the AzureML run are compared against files in `expected_folder`. If neither `run` nor `actual_folder`
+    are provided, a ValueError is raised.
```

Comment:
We should improve the docstring syntax for Sphinx. This is written in Markdown style.

**Current**

```
    Compares a set of files in a folder, against files in either the other folder or files stored in the given
    AzureML run. Each file that is present in the "expected" folder must be also present in the "actual" folder
    (or the AzureML run), with exactly the same contents, in the same folder structure.
    For example, if there is a file "<expected>/foo/bar/contents.txt", then there must also be a file
    "<actual>/foo/bar/contents.txt". If `actual_folder` is provided, then this is used to compare files against the
    set file files in `expected_folder`, irrespective of the value in `run`. If `run` is provided, the files
    uploaded to the AzureML run are compared against files in `expected_folder`. If neither `run` nor `actual_folder`
    are provided, a ValueError is raised.
```

![Web capture_5-4-2022_104253_wsl%24](https://user-images.githubusercontent.com/12688084/161726775-e78c83f1-982b-4245-a311-2f01036b200d.jpeg)

**Slightly better**

```
    Compares a set of files in a folder, against files in either the other folder or files stored in the given
    AzureML run. Each file that is present in the ``expected_folder`` must be also present in the ``actual_folder``
    (or the AzureML run), with exactly the same contents, in the same folder structure.
    For example, if there is a file ``<expected>/foo/bar/contents.txt``, then there must also be a file
    ``<actual>/foo/bar/contents.txt``. If ``actual_folder`` is provided, then this is used to compare files against the
    set file files in ``expected_folder``, irrespective of the value in ``run``. If ``run`` is provided, the files
    uploaded to the AzureML run are compared against files in ``expected_folder``. If neither ``run`` nor `actual_folder`
    are provided, a :exc:`ValueError` is raised.

    :param expected_folder: A folder with files that are expected to be present.
    :param actual_folder: The output folder with the actually produced files.
    :param run: An AzureML run
    :param csv_relative_tolerance: When comparing CSV files, use this as the maximum allowed relative discrepancy.
    If 0.0, do not allow any discrepancy.
    :return: A list of human readable error messages, with message and file path. If no errors are found, the list is
    empty.
```

![Web capture_5-4-2022_105149_wsl%24](https://user-images.githubusercontent.com/12688084/161728206-ce195422-67c9-4ad1-8848-a3606104c27d.jpeg)

Reply:
Even more important when other classes or methods in the library are referenced in the docstring.

Reply:
See [a TorchIO class](https://torchio.readthedocs.io/transforms/augmentation.html#randomelasticdeformation) for heavy usage of rst syntax for rich rendering.